### PR TITLE
Add a StateExactCirculatingSupply command

### PR DESCRIPTION
**Pull request by arajasek**
At height 136,374, circ supply is 15579755.735329679830248779 FIL

### CODE REVIEW

File: documentation/en/api-methods.md [hunk 1]
```diff
@@ -3035,6 +3036,29 @@ Response:
 }
 ```
 
+### StateExactCirculatingSupply
+TODO: Remove StateCirculatingSupply maybe?
```
Comment (arajasek) [inline documentation/en/api-methods.md:3040]: ew this todo

File: cli/state.go [hunk 2]
```diff
@@ -1638,6 +1639,34 @@ var stateCircSupplyCmd = &cli.Command{
 	},
 }
 
+// TODO: Remove stateCircSupplyCmd
```
Comment (magik6k) [inline cli/state.go:1642]: I'd
* Make this the default output of `circulating-supply`
* Add a `--vm-supply` flag to get the `api.StateCirculatingSupply` one

**Review by magik6k — COMMENTED** [review]
Logic looks good

File: cli/state.go (hunk 2 above)
Comment (arajasek) [reply cli/state.go:1642]: Yeah, sounds like the approx CircSupply isn't going anywhere, so I will do this

File: api/api_full.go [hunk 3]
```diff
@@ -400,6 +400,9 @@ type FullNode interface {
 
 	// StateCirculatingSupply returns the circulating supply of Filecoin at the given tipset
 	StateCirculatingSupply(context.Context, types.TipSetKey) (CirculatingSupply, error)
+	// TODO: Remove StateCirculatingSupply maybe?
```
[the remark this answers is not in the document]
Comment (magik6k) [reply api/api_full.go:403]: *StateExactCirculatingSupply

File: chain/stmgr/stmgr.go [hunk 4]
```diff
@@ -1257,6 +1263,91 @@ func (sm *StateManager) GetCirculatingSupply(ctx context.Context, height abi.Cha
 	return csi.FilCirculating, nil
 }
 
+func (sm *StateManager) GetExactCirculatingSupply(ctx context.Context, height abi.ChainEpoch, st *state.StateTree) (abi.TokenAmount, error) {
+	circ := big.Zero()
+	unCirc := big.Zero()
+	err := st.ForEach(func(a address.Address, actor *types.Actor) error {
```
Comment (whyrusleeping) [inline chain/stmgr/stmgr.go:1269]: should skip further processing on actors with zero balance

Comment (arajasek) [reply chain/stmgr/stmgr.go:1269]: ack

File: chain/stmgr/stmgr.go [hunk 5]
```diff
@@ -1357,12 +1362,91 @@ func (sm *StateManager) GetCirculatingSupplyDetailed(ctx context.Context, height
...
+			break
+		case a == _init.Address ||
+			a == reward.Address ||
+			a == verifreg.Address ||
+			// The power actor itself should never receive funds
+			a == power.Address ||
+			a == builtin.SystemActorAddr ||
+			a == builtin.CronActorAddr ||
+			a == builtin.BurntFundsActorAddr ||
+			a == builtin.SaftAddress ||
```
Comment (arajasek) [inline chain/stmgr/stmgr.go:1380]: We could start including funds in this address post-liftoff, but eh

**Review by raulk — COMMENTED** [review]
Not being completely in the know about the rationale behind these changes (the PR description doesn't explain much, which makes it hard to follow), it looks like a smell that:

1. we have two circulating supplies,
2. we're choosing to expose the "exact" one via the API by default, but letting the VM use the _non-exact_ value.

Intuitively, something feels a bit off here, either the divergence in values, or the naming. The circulating supply is a crucial cryptoecon variable, and IMO it should be as minimally ambiguous as possible.

**Comment by arajasek** [comment]
@raulk You're not wrong, as I understand it, we're kinda stuck between two requirements here. 

- We must have an accurate calculation of the liquid FIL in the network. This is necessarily slow, but is the correct number to report to anyone external who wants to know how much FIL is circulating.
- We must have a quick approximation of the liquid FIL in the network. This is needed to be used internally when doing CE math (and so by the VM).

The naming is intended to reflect that external parties probably wanna use the slow, accurate method. Very open to hearing recommendations, though.

**Comment by raulk** [comment]
@arajasek 

> This is needed to be used internally when doing CE math (and so by the VM).

Apologies for ignorance, what does CE mean here?

File: chain/stmgr/stmgr.go (hunk 5 above)
Comment (arajasek) [reply chain/stmgr/stmgr.go:1380]: or just include it now, there's only 4M in it.

**Comment by arajasek** [comment]
> @arajasek
> 
> > This is needed to be used internally when doing CE math (and so by the VM).
> 
> Apologies for ignorance, what does CE mean here?

CryptoEcon, sorry

**Review by whyrusleeping — APPROVED** [review]
